lmdeploy/pytorch_poc/patch/baichuan.py

```diff
@@ -0,0 +1,327 @@
+                query_states, key_states, cos, sin, position_ids)
+
+        kv_seq_length = position_ids[..., -1] + 1
+        q_seq_length = kv_seq_length - kv_seq_length.new_tensor(
+            history_lengths)
+        q_start_loc = q_seq_length.cumsum(0)
+        q_start_loc = torch.cat([q_start_loc.new_zeros(1), q_start_loc[:-1]])
+        context.fill_cache(
+            key_states,
+            value_states,
+            q_start_loc,
+            q_seq_length,
+            past_key_value[0],
+            past_key_value[1],
+        )
+        attn_output = torch.empty_like(query_states)
+
+        block_offsets = context.block_offsets
+        block_size = past_key_value[0].size(1)
+        paged_attention_fwd(query_states,
+                            past_key_value[0],
+                            past_key_value[1],
+                            attn_output,
+                            block_offsets,
+                            b_start_loc=q_start_loc,
+                            b_seq_len=q_seq_length,
+                            b_kv_seq_len=kv_seq_length,
+                            max_input_len=max_seq_len,
+                            BLOCK=block_size)
+        attn_output = attn_output.reshape(-1, origin_self.hidden_size)
+
+        attn_output = origin_self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
+
+        return attn_output, attn_weights, past_key_value
+
+
+class BaichuanLayer(torch.nn.Module):
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        output_attentions: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+    ) -> Tuple[torch.FloatTensor, Optional[Tuple[torch.FloatTensor,
+                                                 torch.FloatTensor]]]:
+
+        residual = hidden_states
+
+        hidden_states = self.origin_mod.input_layernorm(hidden_states)
+
+        # Self Attention
+        hidden_states, self_attn_weights, present_key_value = self.origin_mod.self_attn(  # noqa
+            hidden_states=hidden_states,
+            position_ids=position_ids,
+            attention_mask=attention_mask,
+            past_key_value=past_key_value,
+            output_attentions=output_attentions,
+            use_cache=use_cache,
+        )
+        hidden_states = residual + hidden_states
+
+        # Fully Connected
+        residual = hidden_states
+        hidden_states = self.origin_mod.post_attention_layernorm(hidden_states)
+        hidden_states = self.origin_mod.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        outputs = (hidden_states, )
+
+        if use_cache:
+            outputs += (present_key_value, )
+
+        return outputs
+
+
+class BaichuanModel(nn.Module):
```

Comment:
What is the different between this rewrite and the one in llama? Can we register LlamaModel rewrite as BaichuanModel?